components/eamxx/src/physics/cosp/eamxx_cosp.cpp

```diff
@@ -0,0 +1,190 @@
+  add_field<Required>("p_int",             scalar3d_layout_int, Pa,     grid_name);
+  //add_field<Required>("height_mid",  scalar3d_layout_mid, m,      grid_name);
+  //add_field<Required>("height_int",  scalar3d_layout_int, m,      grid_name);
+  add_field<Required>("T_mid",            scalar3d_layout_mid, K,      grid_name);
+  add_field<Required>("qv",               scalar3d_layout_mid, Q,      grid_name, "tracers");
+  add_field<Required>("qc",               scalar3d_layout_mid, Q,      grid_name, "tracers");
+  add_field<Required>("qi",               scalar3d_layout_mid, Q,      grid_name, "tracers");
+  add_field<Required>("cldfrac_tot_for_analysis", scalar3d_layout_mid, nondim, grid_name);
+  // Optical properties, should be computed in radiation interface
+  add_field<Required>("dtau067",     scalar3d_layout_mid, nondim, grid_name); // 0.67 micron optical depth
+  add_field<Required>("dtau105",     scalar3d_layout_mid, nondim, grid_name); // 10.5 micron optical depth
+  // Effective radii, should be computed in either microphysics or radiation interface
+  // TODO: should these be meters or microns? Was meters before, but using "m" instead
+  // of "micron" seemed to cause prim_model_finalize to throw error with the following:
+  // ABORTING WITH ERROR: Error! prim_init_model_f90 was not called yet (or prim_finalize_f90 was already called).
+  // P3 defines this field with micron instead of meters units, so is this a unit conversion issue?
+  add_field<Required>("eff_radius_qc",     scalar3d_layout_mid, micron,      grid_name);
+  add_field<Required>("eff_radius_qi",     scalar3d_layout_mid, micron,      grid_name);
+  // Set of fields used strictly as output
+  add_field<Computed>("isccp_cldtot", scalar2d_layout, nondim, grid_name);
+  add_field<Computed>("isccp_ctptau", scalar4d_layout_ctptau, nondim, grid_name, 1);
+  add_field<Computed>("isccp_mask"  , scalar2d_layout, nondim, grid_name);
+}
+
+// =========================================================================================
+void Cosp::initialize_impl (const RunType /* run_type */)
+{
+  // Determine how often to call COSP, specified as number of atm steps (>0) or number of hours (<0)
+  m_cosp_frequency = m_params.get<Int>("cosp_frequency", 1);
+
+  // How many subcolumns to use for COSP
+  m_num_subcols = m_params.get<Int>("cosp_subcolumns", 10);
+
+  // Set property checks for fields in this process
+  CospFunc::initialize(m_num_cols, m_num_subcols, m_num_levs);
+}
+
+// =========================================================================================
+void Cosp::run_impl (const double dt)
+{
+  // Determine if we should update COSP this timestep; use rad function to compare COSP frequency with this timestep
+  auto ts = timestamp();
+  auto cosp_freq_in_steps = m_cosp_frequency;
+  if (m_cosp_frequency < 0) {
+      // interpret cosp_freq as nhours
+      cosp_freq_in_steps = -3600.0 * m_cosp_frequency / dt;
```

Comment:
I would double check that `dt` divides 3600 here, or else you may get some unexpected behavior. I'm sure this is (currently) always the case, but just in case someone accidentally (or willingly) changed, say, some subcycling parameter, better be safe than sorry.

Reply:
Added an EKAT_REQUIRE_MSG here for this.